File: lib/services/printer.dart
import "dart:io";
import "dart:typed_data";

import "package:flutter/foundation.dart";
import "package:printing/printing.dart";
import "package:qbitsinc_printer_manager/services/ticket.dart";
import "package:qbitsinc_printer_manager/store/app_main_store.dart";

Future<void> printTicket(
  String destination,
  dynamic data,
  int printerSize,
) async {
  AppMainStore.instance.update(
    AppMainState(
      status: AppStatus.printing,
      message: "Printing Document to ${destination}",
    ),
  );
  Uint8List contentPDF = await generatePDFReceipt(data, printerSize);
  try {
    Printer printer = await Printing.listPrinters()
        .asStream()
        .map(
          (printers) =>
              printers.firstWhere((printer) => printer.name == destination),
        )
        .first;
    await Printing.directPrintPdf(
      printer: printer,
      onLayout: (_) => contentPDF,
    );
    AppMainStore.instance.update(AppMainState(status: AppStatus.connected));
  } catch (e) {
    if (e is StateError) {
      throw Exception("Printer: $destination, not found: $e");
    }
    throw Exception("Error sending ticket to printer at $destination: $e");
  }
}

Future<void> printNetworkTicket(
  String destination,
  dynamic data,
  int printerSize,
) async {
  try {
    // 1. Build the receipt data using the same logic as local printing.
    var receipt = await generatePDFReceipt(data, printerSize);
    Uint8List receiptUintList = Uint8List.fromList(receipt);

    // 2. Connect to the network printer via raw TCP socket on port 9100.
    final socket = await Socket.connect(
      destination,
      9100,
      timeout: const Duration(seconds: 5),
    );

    // 3. Send the data.
    socket.add(receiptUintList);
    await socket.flush();

    // 4. Close the socket.
    await socket.close();
  } catch (e) {
    throw Exception(
      "Error sending ticket to network printer at $destination: $e",
    );
  }
}

Future<void> printInvoice(
  dynamic data,
  String destination, [
  int printerSize = 58,
]) async {
  print("Printing invoice to $destination and data: $data");
  // Use a more robust regex to check for an IPv4 address format.
  if (RegExp(r"^\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3}$").hasMatch(destination)) {
    await printNetworkTicket(destination, data, printerSize);
  } else {
    await printTicket(destination, data, printerSize);
  }
}

File: server/core/services/printer.py
import re
import socket
from typing import Any

from core.constants.doc import TMP_DIR
from core.services.printers import PrintersService
from core.services.ticket import ReceiptType, generate_pdf_receipt
from core.state.app_state import AppStateObserver, AppStateType


async def print_ticket(
    destination: str, data: dict[str, Any], printer_size: int, doc_type: str
) -> None:
    """Prints a ticket to a local CUPS printer."""
    try:
        receipt_type = (
            ReceiptType.COMMAND
            if re.match("^comm", doc_type, re.IGNORECASE)
            else ReceiptType.RECEIPT
        )
        content_pdf = await generate_pdf_receipt(data, printer_size, receipt_type)

        AppStateObserver.add(
            AppStateType.PRINTING, message=f"Printing ticket to: {destination}"
        )
        with open(TMP_DIR / "receipt.pdf", "wb") as f:
            f.write(content_pdf)
            PrintersService().sendToPrint(
                str((TMP_DIR / "receipt.pdf").resolve()), destination
            )
            AppStateObserver.add(AppStateType.CONNECTED)

    except Exception as e:
        AppStateObserver.add(
            AppStateType.ERROR,
            message=f"Error sending ticket to printer at {destination}: {e}",
        )


async def print_network_ticket(
    destination: str, data: dict[str, Any], printer_size: int, doc_type: str
) -> None:
    """Prints a ticket to a network printer."""
    try:
        receipt_type = (
            ReceiptType.COMMAND
            if re.match("^comm", doc_type, re.IGNORECASE)
            else ReceiptType.RECEIPT
        )
        receipt_pdf = await generate_pdf_receipt(data, printer_size, receipt_type)
        AppStateObserver.add(
            AppStateType.PRINTING, message=f"Printing {receipt_type} to: {destination}"
        )
        with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as s:
            s.settimeout(5)
            s.connect((destination, 9100))
            s.sendall(receipt_pdf)
            AppStateObserver.add(AppStateType.CONNECTED)
    except Exception as e:
        AppStateObserver.add(
            AppStateType.ERROR,
            message=f"Error sending ticket to printer at {destination}: {e}",
        )


async def print_invoice(
    data: dict[str, Any], destination: str, doc_type: str, printer_size: int = 58
) -> None:
    """Determines whether to print to a local or network printer."""
    if re.match(r"^\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3}$", destination):
        await print_network_ticket(destination, data, printer_size, doc_type)
    else:
        await print_ticket(destination, data, printer_size, doc_type)
